File: ts/kip-do/src/exposure.ts
/**
 * The exposure log (Spec §66.8).
 *
 * A read never reinforces memory (§2.13), yet a Brain needs to know what it
 * retrieved and used. The log keeps that signal without turning recall into a
 * write path:
 *
 * ```text
 * not cognitive state   no element, no space_seq, no Change Envelope entry
 * never evidence        never cited, never corroboration, never confidence
 * governed              read only under read_audit; an element the reader may
 *                       not discover is omitted, never counted
 * bounded               erased with its element (§60.7)
 * append-only           entries are never rewritten; the host records them
 * ```
 *
 * Mirrors `rs/anda_cognitive_nexus/src/exposure.rs`.
 */

import type { AuthContext, EffectiveAuthority } from './governance/index.js'
import { requirePermitted, spaceResource } from './governance/index.js'
import { errors } from './errors.js'
import type { JsonMap } from './json.js'
import { formatElementId, tryParseElementId } from './id.js'
import { State, type Element } from './store/index.js'
import type { Store } from './store/store.js'
import { nowTime } from './time.js'

/** The most entries one call records or returns. */
export const MAX_EXPOSURES = 1000

export type Exposure = 'retrieved' | 'used'

/** One exposure a host records; the Space, time and Principal are the engine's. */
export interface ExposureInput {
  element_id: string
  exposure: Exposure
  /** The snapshot sequence of the read that exposed the element. */
  snapshot_seq: number
  /** The decision that used it; required for `used`. */
  decision_ref?: string
  /** The recall that returned it. */
  recall_ref?: string
}

/** One entry of the log (`kip-cognitive-records.schema.json#/$defs/ExposureRecord`). */
export interface ExposureRecord {
  space_id: string
  element_id: string
  exposure: Exposure
  snapshot_seq: number
  recorded_at: string
  principal_id: string
  decision_ref?: string
  recall_ref?: string
}

/** A bounded page of the log, oldest first. */
export interface ExposureQuery {
  element_id?: string
  /** The `next_cursor` of an earlier page. */
  cursor?: string
  /** `1..=1000`; default 100. */
  limit?: number
}

interface ExposureRow {
  id: number
  space: string
  element: string
  exposure: string
  snapshot_seq: number
  recorded_at: string
  principal_id: string
  decision_ref: string
  recall_ref: string
}

export interface ExposureHost {
  store: Store
  authority: EffectiveAuthority
  auth: AuthContext
}

function readable(host: ExposureHost, space: string, ref: string): Element {
  const unavailable = () => errors.notFoundOrNotVisible(`${ref} is unavailable`)
  const id = tryParseElementId(ref)
  const element = id === null ? null : host.store.load(id)
  if (!element || element.row.space !== space ||
    element.row.state === State.PURGED || element.row.state === State.PENDING) {
    throw unavailable()
  }
  if (!host.authority.mayRead(element, host.auth)?.content) throw unavailable()
  return element
}

/** Appends entries; they take no Space sequence and change no element. */
export function recordExposures(host: ExposureHost, space: string, entries: ExposureInput[]): JsonMap {
  if (!Array.isArray(entries) || entries.length === 0 || entries.length > MAX_EXPOSURES) {
    throw errors.constraintViolation(`record 1..=${MAX_EXPOSURES} exposure entries at a time`)
  }
  const current = host.store.space(space)?.seq ?? 0
  const recordedAt = nowTime()
  const rows: Omit<ExposureRow, 'id'>[] = []
  for (const entry of entries) {
    if (entry.exposure !== 'retrieved' && entry.exposure !== 'used') {
      throw errors.constraintViolation('exposure is retrieved or used')
    }
    const element = readable(host, space, entry.element_id)
    if (!Number.isSafeInteger(entry.snapshot_seq) || entry.snapshot_seq < 0 || entry.snapshot_seq > current) {
      throw errors.constraintViolation(`snapshot_seq ${entry.snapshot_seq} is ahead of the Space at ${current}`)
    }
    let decision = ''
    if (entry.decision_ref !== undefined) {
      if (readable(host, space, entry.decision_ref).kind !== 'Activity') {
        throw errors.constraintViolation('decision_ref names the Activity that recorded the decision')
      }
      decision = entry.decision_ref
    } else if (entry.exposure === 'used') {
      throw errors.constraintViolation('a used exposure names the decision that used the element')
    }
    const recall = entry.recall_ref ?? ''
    if ([...recall].length > 1024) throw errors.constraintViolation('recall_ref is at most 1024 characters')
    rows.push({
      space,
      element: formatElementId({ kind: element.kind, seq: element.row.id }),
      exposure: entry.exposure,
      snapshot_seq: entry.snapshot_seq,
      recorded_at: recordedAt,
      principal_id: host.auth.principal_id,
      decision_ref: decision,
      recall_ref: recall,
    })
  }
  for (const row of rows) {
    host.store.sql.exec(
      'INSERT INTO kip_exposures (space, element, exposure, snapshot_seq, recorded_at, principal_id, decision_ref, recall_ref) VALUES (?, ?, ?, ?, ?, ?, ?, ?)',
      row.space, row.element, row.exposure, row.snapshot_seq, row.recorded_at, row.principal_id, row.decision_ref, row.recall_ref,
    )
  }
  return { recorded: rows.length }
}

/** Reads the log under `read_audit`, omitting elements the reader may not discover. */
export function readExposures(host: ExposureHost, space: string, query: ExposureQuery = {}): JsonMap {
  const limit = query.limit ?? 100
  if (!Number.isSafeInteger(limit) || limit < 1 || limit > MAX_EXPOSURES) {
    throw errors.constraintViolation(`exposure page limit is 1..=${MAX_EXPOSURES}`)
  }
  let after = 0
  if (query.cursor !== undefined) {
    const match = /^exposure:(\d+)$/.exec(query.cursor)
    if (!match) throw errors.cursorInvalid('invalid exposure cursor')
    after = Number(match[1])
  }
  requirePermitted(host.authority.authorize('read_audit', spaceResource(), host.auth))
  const rows = (query.element_id === undefined
    ? host.store.sql.exec('SELECT * FROM kip_exposures WHERE space = ? AND id > ? ORDER BY id', space, after)
    : host.store.sql.exec('SELECT * FROM kip_exposures WHERE space = ? AND element = ? AND id > ? ORDER BY id', space, query.element_id, after)
  ).toArray() as unknown as ExposureRow[]
  const records: ExposureRecord[] = []
  let last: number | null = null
  let more = false
  for (const row of rows) {
    if (records.length === limit) { more = true; break }
    last = row.id
    const id = tryParseElementId(row.element)
    const element = id === null ? null : host.store.load(id)
    if (!element || element.row.state === State.PURGED || !host.authority.mayRead(element, host.auth)) continue
    records.push({
      space_id: row.space,
      element_id: row.element,
      exposure: row.exposure === 'used' ? 'used' : 'retrieved',
      snapshot_seq: row.snapshot_seq,
      recorded_at: row.recorded_at,
      principal_id: row.principal_id,
      ...(row.decision_ref ? { decision_ref: row.decision_ref } : {}),
      ...(row.recall_ref ? { recall_ref: row.recall_ref } : {}),
    })
  }
  return {
    records: records as unknown as JsonMap[],
    next_cursor: more && last !== null ? `exposure:${last}` : null,
  }
}

/** Removes an erased element's entries (§60.7). */
export function removeExposures(store: Store, space: string, element: string): void {
  store.sql.exec('DELETE FROM kip_exposures WHERE space = ? AND element = ?', space, element)
}

/** How many entries an element still has, for erasure verification. */
export function exposureCount(store: Store, space: string, element: string): number {
  return Number(
    (store.sql.exec('SELECT COUNT(*) AS n FROM kip_exposures WHERE space = ? AND element = ?', space, element).one() as { n: number }).n,
  )
}

File: rs/anda_cognitive_nexus/src/exposure.rs
//! # The exposure log (Spec §66.8)
//!
//! A read never reinforces memory (§2.13), yet a Brain needs to know what it
//! retrieved and used: use is the strongest signal that a memory should stay
//! accessible. The exposure log keeps that signal without turning recall into
//! a write path:
//!
//! ```text
//! not cognitive state   no element, no space_seq, no Change Envelope entry
//! never evidence        never cited, never corroboration, never confidence
//! governed              read only under read_audit; an element the reader may
//!                       not discover is omitted, never counted
//! bounded               erased with its element (§60.7)
//! append-only           entries are never rewritten; the host records them
//! ```
//!
//! The host or Memory Interface Adapter records `retrieved` for items it
//! returned and `used` for a DecisionRecord's `used_refs`; Maintenance reads
//! the log in bounded batches and writes any reinforcement explicitly
//! (§59.1). Nothing here changes confidence, strength or utility.

use crate::{
    error::db_error,
    governance::{AuthContext, EffectiveAuthority, Permission, ResourceContext},
    id::ElementId,
    nexus::Session,
    store::{Element, Store, eq_field, rows::ExposureRow, rows::state},
};
use anda_db::query::{Filter, RangeQuery};
use anda_db_schema::Fv;
use anda_kip::{
    Json, KipError,
    cognitive::{Exposure, ExposureRecord},
};
use serde::{Deserialize, Serialize};
use serde_json::json;

/// The most entries one call records or returns.
pub const MAX_EXPOSURES: usize = 1000;
const READ_BATCH: usize = 256;

/// One exposure a host records. The Space, the time and the Principal are the
/// engine's, never the caller's.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ExposureInput {
    pub element_id: String,
    pub exposure: Exposure,
    /// The snapshot sequence of the read that exposed the element.
    pub snapshot_seq: u64,
    /// The decision that used it; required for `used`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub decision_ref: Option<String>,
    /// The recall that returned it.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recall_ref: Option<String>,
}

/// A bounded page of the log, oldest first.
#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ExposureQuery {
    /// Only this element's entries.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub element_id: Option<String>,
    /// The `next_cursor` of an earlier page.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cursor: Option<String>,
    /// `1..=1000`; default 100.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<usize>,
}

impl Store {
    /// Removes every exposure entry of one element; part of its erasure.
    pub(crate) async fn remove_exposures(&self, element: &str) -> Result<(), KipError> {
        let table = self.exposures();
        let ids = table
            .query_all_ids(eq_field("element", Fv::Text(element.into())))
            .await
            .map_err(db_error)?;
        for id in ids {
            table.remove(id).await.map_err(db_error)?;
        }
        Ok(())
    }

    /// How many exposure entries one element still has.
    pub(crate) async fn exposure_count(&self, element: &str) -> Result<usize, KipError> {
        Ok(self
            .exposures()
            .query_all_ids(eq_field("element", Fv::Text(element.into())))
            .await
            .map_err(db_error)?
            .len())
    }
}

impl Session {
    /// Appends exposure entries (§66.8).
    ///
    /// The caller must be able to read each element it reports, and a `used`
    /// entry names the decision that used it — an Activity the caller can
    /// read. The entries take no Space sequence and change no element.
    pub async fn record_exposures(
        &self,
        space: &str,
        entries: Vec<ExposureInput>,
    ) -> Result<Json, KipError> {
        if entries.is_empty() || entries.len() > MAX_EXPOSURES {
            return Err(KipError::constraint_violation(format!(
                "record 1..={MAX_EXPOSURES} exposure entries at a time"
            )));
        }
        self.with_authority(space, async |authority| {
            let store = &self.nexus.store;
            let current = store.get_space(space).await?.seq;
            let recorded_at = crate::time::now();
            let mut rows = Vec::with_capacity(entries.len());
            for entry in &entries {
                let element =
                    readable(store, space, &entry.element_id, &authority, &self.auth).await?;
                if entry.snapshot_seq > current {
                    return Err(KipError::constraint_violation(format!(
                        "snapshot_seq {} is ahead of the Space at {current}",
                        entry.snapshot_seq
                    )));
                }
                let decision = match (&entry.decision_ref, entry.exposure) {
                    (Some(reference), _) => {
                        let decision =
                            readable(store, space, reference, &authority, &self.auth).await?;
                        if !matches!(decision, Element::Activity(_)) {
                            return Err(KipError::constraint_violation(
                                "decision_ref names the Activity that recorded the decision",
                            ));
                        }
                        reference.clone()
                    }
                    (None, Exposure::Used) => {
                        return Err(KipError::constraint_violation(
                            "a used exposure names the decision that used the element",
                        ));
                    }
                    (None, Exposure::Retrieved) => String::new(),
                };
                let recall = entry.recall_ref.clone().unwrap_or_default();
                if recall.chars().count() > 1024 {
                    return Err(KipError::constraint_violation(
                        "recall_ref is at most 1024 characters",
                    ));
                }
                rows.push(ExposureRow {
                    _id: 0,
                    space: space.to_string(),
                    element: element.id().to_string(),
                    exposure: match entry.exposure {
                        Exposure::Retrieved => "retrieved".into(),
                        Exposure::Used => "used".into(),
                    },
                    snapshot_seq: entry.snapshot_seq,
                    recorded_at: recorded_at.clone(),
                    principal_id: self.auth.principal_id.clone(),
                    decision_ref: decision,
                    recall_ref: recall,
                });
            }
            let table = store.exposures();
            for row in &rows {
                table.add_from(row).await.map_err(db_error)?;
            }
            table.flush(crate::tx::now_ms()).await.map_err(db_error)?;
            Ok(json!({"recorded": rows.len()}))
        })
        .await
    }

    /// Reads the exposure log, oldest first (§66.8).
    ///
    /// Requires `read_audit`. An entry whose element the reader may not
    /// discover is omitted, and the page says nothing about how many were.
    pub async fn read_exposures(
        &self,
        space: &str,
        query: ExposureQuery,
    ) -> Result<Json, KipError> {
        let limit = query.limit.unwrap_or(100);
        if limit == 0 || limit > MAX_EXPOSURES {
            return Err(KipError::constraint_violation(format!(
                "exposure page limit is 1..={MAX_EXPOSURES}"
            )));
        }
        let after = match &query.cursor {
            None => 0,
            Some(cursor) => cursor
                .strip_prefix("exposure:")
                .and_then(|n| n.parse::<u64>().ok())
                .ok_or_else(|| {
                    KipError::new(
                        anda_kip::KipErrorCode::CursorInvalid,
                        "invalid exposure cursor",
                    )
                })?,
        };
        let _guard = self.nexus.read_guard().await?;
        let authority = self.effective_authority(space).await?;
        authority
            .authorize(
                Permission::ReadAudit,
                &ResourceContext::default(),
                &self.auth,
            )
            .into_result()?;
        let store = &self.nexus.store;
        let table = store.exposures();
        let mut scanned = after;
        let mut records = Vec::new();
        let mut last = None;
        let mut more = false;
        'pages: loop {
            // Scan the id range directly: combining multiple equality indexes
            // would materialize their full intersection before applying a limit.
            // Each batch starts after the last scanned id, including hidden rows.
            let ids = table
                .query_ids(
                    Filter::Field(("_id".into(), RangeQuery::Gt(Fv::U64(scanned)))),
                    Some(READ_BATCH),
                )
                .await
                .map_err(db_error)?;
            let exhausted = ids.len() < READ_BATCH;
            for id in ids {
                scanned = id;
                let row: ExposureRow = table.get_as(id).await.map_err(db_error)?;
                if row.space != space
                    || query
                        .element_id
                        .as_ref()
                        .is_some_and(|element| *element != row.element)
                {
                    continue;
                }
                let discoverable = match row.element.parse::<ElementId>() {
                    Ok(element) => store
                        .get_element(element)
                        .await
                        .ok()
                        .is_some_and(|element| {
                            element.state() != state::PURGED
                                && authority.may_read(&element, &self.auth).is_some()
                        }),
                    Err(_) => false,
                };
                if !discoverable {
                    continue;
                }
                // Look ahead to a visible entry, so hidden trailing rows never
                // produce a cursor or an extra empty page.
                if records.len() == limit {
                    more = true;
                    break 'pages;
                }
                last = Some(id);
                records.push(ExposureRecord {
                    space_id: row.space,
                    element_id: row.element,
                    exposure: if row.exposure == "used" {
                        Exposure::Used
                    } else {
                        Exposure::Retrieved
                    },
                    snapshot_seq: row.snapshot_seq,
                    recorded_at: row.recorded_at,
                    principal_id: row.principal_id,
                    decision_ref: Some(row.decision_ref).filter(|r| !r.is_empty()),
                    recall_ref: Some(row.recall_ref).filter(|r| !r.is_empty()),
                });
            }
            if exhausted {
                break;
            }
        }
        Ok(json!({
            "records": records,
            "next_cursor": if more { last.map(|id| format!("exposure:{id}")) } else { None },
        }))
    }
}

/// An element the caller may read, existence-neutrally.
async fn readable(
    store: &Store,
    space: &str,
    reference: &str,
    authority: &EffectiveAuthority,
    auth: &AuthContext,
) -> Result<Element, KipError> {
    let unavailable = || KipError::not_found_or_not_visible(format!("{reference} is unavailable"));
    let id: ElementId = reference.parse().map_err(|_| unavailable())?;
    let element = store.get_element(id).await.map_err(|_| unavailable())?;
    if element.space() != space
        || matches!(element.state(), state::PURGED | state::PENDING)
        || !authority
            .may_read(&element, auth)
            .is_some_and(|v| v.content)
    {
        return Err(unavailable());
    }
    Ok(element)
}
